ARCs/arc-0060.md

```diff
@@ -0,0 +1,449 @@
+---
+arc: 60
+title: Algorand Wallet Structured Data Signing API
+description: API function for signing structured data
+author: Stefano De Angelis (@deanstef)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/284
+status: Draft
+type: Standards Track
+category: Interface
+created: 2024-02-28
+requires: 1
+---
+
+# Algorand Wallet Structured Data Signing API
+
+> This ARC is inspired by [ARC-1](./arc-0001.md).
+
+## Abstract
+
+ARC-1 defines a standard for signing Algorand transactions that are represented as structured objects. This proposal extends the signing process to byte arrays encoded with a standardized structure.
+
+[ARC-60](./arc-0060.md) defines an API for wallets to sign structured data that are not Algorand transactions.
+
+## Motivation
+
+Signing data is a common and critical operation. Users may need to sign data for multiple reasons (e.g. delegate signatures, DIDs, signing documents, authentication).
+
+Algorand wallets need a standard approach to byte signing to unlock self-custodial services and protect users from malicious and attack-prone signing workflows.
+
+This ARC provides a standard API for bytes signing. The API encodes byte arrays to be signed into well-structured JSON schemas together with additional metadata. It requires wallets to validate the signing inputs, notify users about what they are signing and warn them in case of dangerous signing requests.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+> Comments like this are non-normative
+
+### Overview
+
+> This section is non-normative
+
+ARC-60 defines a function `signData(signingData, metadata)` for signing data.
+
+`signingData` is a `StdSigData` object composed of the signing `data` that instantiates a known JSON Schema and the `signer`'s public key.
+
+`metadata` is a `StdSignMetadata` object that describes the signature `scope`, the JSON `schema`, and the `encoding` used to represent the signing bytes. The JSON schema takes two elements: the `bytes` being signed and an optional `prefix`.
+
+Ultimately, the function computes the message to be signed as `msg=(<prefix> || <data.bytes>)` and returns the signature `Sig(msg, sk)` where `sk` is the `signer`'s signing key.
+
+There are two possible use cases for `signData`:
+
+1. Sign data with one signer that the wallet knows. For example:
+
+```json
+{
+  "data": "/{.../}",
+  "signer": "xxxx"
+}
+{
+  "scope": "...",
+  "schema": "/{.../}",
+  "encoding": "..."
+}
```

Comment:
IRL we agree that both`data` and `schema` can be JSON objects rather than strings

https://github.com/algorandfoundation/ARCs/pull/284/commits/43a7178ae4a37d71eb78abf7cd4049f40ea0d12a